[BEAM-3182] Hookup Nexmark query models to SQL

apilloud opened the pull request:
This change makes SQL queries output Nexmark model datatypes, then hooks them up to the model verification test.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [X] Write a pull request description that is detailed enough to understand:
   - [X] What the pull request does
   - [X] Why it does it
   - [X] How it does it
   - [X] Why this approach
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.

apilloud commented:
R: @akedin

apilloud commented:
run python precommit

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/SqlQuery1.java:
```diff
@@ -79,4 +82,21 @@ public SqlQuery1() {
     return ADAPTERS.get(Bid.class).getRowType().getRowCoder();
   }
+
+  public PCollection<? extends KnownSize> applyModel(PCollection<Row> rows) {
```
would it make more sense to keep all the mapping from and to models in a single place, e.g. [`ModelAdaptersMapping`](https://github.com/apache/beam/blob/11ad97e42015dc497fb5cc3d38090e64f9c7418b/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelAdaptersMapping.java), by adding a method to [`ModelFieldsAdapter`](https://github.com/apache/beam/blob/11ad97e42015dc497fb5cc3d38090e64f9c7418b/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapter.java), and maybe a utility method in [`ToRow`](https://github.com/apache/beam/blob/a13c7d9177c34017955cbfbfc29a7f4a698047c1/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/ToRow.java)

apilloud replied:
I think it does make sense to add the reverse mappings to `ModelAdaptersMapping` and referenece that here. I'll update.

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/NexmarkSqlQuery.java:
```diff
@@ -46,9 +44,8 @@ public NexmarkSqlQuery(NexmarkConfiguration configuration,
-        .apply(RowSize.parDo())
-        .setCoder(RowSize.CODER);
+    PCollection<? extends KnownSize> resultRecordsSizes =
+        queryTransform.applyModel(queryResults);
```
There's only about 3 pojo models we're interested in, so if you make `RowSize.parDo()` convert to those models  instead of converting to a custom `RowSize` (see my other comment), then you don't need to change the interfaces or inheritance

lukecwik commented:
@echauchot As a committer, can you take a look at this?

kennknowles commented:
Etienne is still off right now, so I am happy to take a look.

lukecwik commented:
Great, I was just looking through open PRs and tried to find people how have some expertise in the space. Wanted to spread the load since the few active committers experience the majority of the load.

apilloud commented:
@akedin I refactored this such that each test specifies its output java class and added the reverse conversion logic to `ModelFieldsAdapter`. I've only ported Query 1, but once all the queries are moved to this model we can remove `RowSize`.

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapter.java:
```diff
@@ -37,4 +43,19 @@ public RowType getRowType() {
+  public abstract T getRowModel(Row row);
+  public abstract Coder getCoder();
+
+  public PCollection<T> applyModel(PCollection<Row> rows) {
```
This is essentially the `PTransform.expand()` method.  

I personally think that it would be clearer if instead of applying the `ParDo`, this method would just return the `PTransform` which would be applied somewhere else, e.g. `ModelFieldsAdapter.toModel(Person.getClass())`. Similar to `Combine.perKey()`, `Window.into()`. This way the pipeline is assembled from `PTransforms` in more declarative style, instead of opaquely modifying the `PCollection`.

And it doesn't feel like all queries should expose `applyModel(PCollection)` just for the sake of `NexmarkSqlQuery` calling it. I think that queries either should explicitly handle the conversion internally (copy-pasting) without exposing this method, or don't handle it at all and leave it to `NexmarkSqlQuery` calling the utility methods.

E.g. I think that `if (outModel == null) {` logic from `NexmarkSqlTransform` can be kept in `NexmarkSqlQuery` or implemented as `ModelFieldsAdapter.toModel()`, without having to introduce the `NexmarkSqlTransform` class and extra

apilloud replied:
I was having difficulty using a `PTransform` without definite types. Might just be my lack of java knowledge. I think once I get all the existing SQL queries moved to this model we can remove this and `NexmarkSqlTransform` and just transform to the java pojo in `expand`. I could do that all in one PR, but it would be rather large.

akedin replied:
Cool, I'm fine with it

echauchot replied:
I agree with what @akedin said about `applyModel`. I like the refactoring you did and the removal of `NexmarkSqlTransform`

apilloud commented:
@kennknowles Please take a look.

apilloud commented:
Ping @kennknowles

kennknowles commented:
Reviewed 2 of 2 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapter.java, line 50 at r2](https://reviewable.io:443/reviews/apache/beam/4718#-L6RpJ1dAHy67lpx44Tu-r2-50:-L6RpysAe6phY2GvkM9e:bj825rw) ([raw file](https://github.com/apache/beam/blob/d7c63d99112b0a19841a6922d39849ae92d74358/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/sql/adapter/ModelFieldsAdapter.java#L50)):*
<details><summary><i>Previously, akedin (Anton Kedin) wrote…</i></summary><blockquote>

Cool, I'm fine with it
</blockquote></details>

I'm also fine with it for now. I think most likely it should all fade away and be implementation details of particular queries, yes?

---

*[sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/NexmarkSqlTransform.java, line 35 at r2](https://reviewable.io:443/reviews/apache/beam/4718#-L6RrHB2z1NbSgwqFL0o:-L6RrHB2z1NbSgwqFL0p:bla6o3x) ([raw file](https://github.com/apache/beam/blob/d7c63d99112b0a19841a6922d39849ae92d74358/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/NexmarkSqlTransform.java#L35)):*
> ```Java
>  * decoupled from Nexmark KnownSize model transform.
>  */
> public abstract class NexmarkSqlTransform extends PTransform<PCollection<Event>, PCollection<Row>> {
> ```

TBH I'm least happy with this class. It appears to be used both as an interface to supply the `applyModel` method as well as a base class to give a default implementation. I would be happier if each query owned its output type, so this PR would be porting SqlQuery0 to simply return the type that is the desired POJO (just like it accepts PCollection<Event> as input, rather than PCollection<Row>).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4718)*
<!-- Sent from Reviewable.io -->

apilloud commented:
@kennknowles I've updated this to go all the way. It's 5x bigger but much smaller then I expected. All the sql queries are still a PTransform, however they return the correct java pojo instead of `Row`.

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/Bid.java:
```diff
@@ -156,6 +157,28 @@ public Bid withoutAnnotation(String annotation) {
+    return auction == other.auction
+        && bidder == other.bidder
+        && price == other.price
+        && dateTime == other.dateTime
```
this made me double check the type of `dateTime` field. I think it's less error-prone to use `Objects.equals()` even for primitive types

akedin commented on sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/model/NameCityStateId.java:
```diff
@@ -87,6 +88,27 @@ public NameCityStateId(String name, String city, String state, long id) {
+    }
+
+    NameCityStateId other = (NameCityStateId) otherObject;
+    return name.equals(other.name)
```
Unless we have [`@DefaultAnnotation(NonNull.class)`](https://github.com/apache/beam/blob/645e1b51e1c83ec4b0b54a774d6615624f3c42ed/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/agg/package-info.java#L22)  in `package-info.java` for this package, then I think we should use `Objects.equals()` for equality checks

apilloud replied:
Fixed. Using Objects.equals() everywhere now.

echauchot commented:
@lukecwik  I'm catching up after sick leave. I'll take a look soon

echauchot commented:
@lukecwik  I'm catching up after sick leave. I'll take a look soon

echauchot commented:
Reviewed 2 of 11 files at r3, 15 of 19 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/NexmarkSqlQuery.java, line 35 at r5](https://reviewable.io:443/reviews/apache/beam/4718#-L6v4xyL0DJ2heCEAxse:-L6v4xyL0DJ2heCEAxsf:b4i5i1o) ([raw file](https://github.com/apache/beam/blob/b8a871c118126a47dcc5d1729e3e3267c8924230/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/NexmarkSqlQuery.java#L35)):*
> ```Java
> public class NexmarkSqlQuery<T extends KnownSize> extends NexmarkQuery {
> 
>   private PTransform<PCollection<Event>, PCollection<T>> queryTransform;
> ```

Indeed, it is better to output pojos than rows

---

*[sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/NexmarkSqlQuery.java, line 45 at r5](https://reviewable.io:443/reviews/apache/beam/4718#-L6vjkG6iVaxyOM9OtLX:-L6vjkG6iVaxyOM9OtLY:b-bpvcmw) ([raw file](https://github.com/apache/beam/blob/b8a871c118126a47dcc5d1729e3e3267c8924230/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/NexmarkSqlQuery.java#L45)):*
> ```Java
>   @Override
>   protected PCollection<KnownSize> applyPrim(PCollection<Event> events) {
>     PCollection<? extends KnownSize> resultRecordsSizes = events.apply(queryTransform);
> ```

I like this refactoring, thanks

---

*[sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/SqlQuery0.java, line 78 at r5](https://reviewable.io:443/reviews/apache/beam/4718#-L6vkioUttxJEA-ckgFs:-L6vkioUttxJEA-ckgFt:b-qv0e8z) ([raw file](https://github.com/apache/beam/blob/b8a871c118126a47dcc5d1729e3e3267c8924230/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/SqlQuery0.java#L78)):*
> ```Java
> 
>     return queryResultsRows
>         .apply(bidParDo())
> ```

Well, it costs  more conversion : pojo -> row -> pojo where before it was only projo -> row but I think the good refactoring is worth this cost

---

*[sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/SqlQuery7.java, line 55 at r5](https://reviewable.io:443/reviews/apache/beam/4718#-L6vn376oAg2p0qvdmEX:-L6vn376oAg2p0qvdmEY:bkmjwvd) ([raw file](https://github.com/apache/beam/blob/b8a871c118126a47dcc5d1729e3e3267c8924230/sdks/java/nexmark/src/main/java/org/apache/beam/sdk/nexmark/queries/sql/SqlQuery7.java#L55)):*
> ```Java
>   private static final String QUERY_TEMPLATE = ""
>       + " SELECT B.auction, B.price, B.bidder, B.dateTime, B.extra "
>       + "    FROM (SELECT B.auction, B.price, B.bidder, B.dateTime, B.extra, "
> ```

Good catch about the missing fields!

---

*[sdks/java/nexmark/src/test/java/org/apache/beam/sdk/nexmark/queries/QueryTest.java, line 98 at r5](https://reviewable.io:443/reviews/apache/beam/4718#-L6vpVrf3UvGIJEdcso7:-L6vpVrf3UvGIJEdcso8:b-pk9h0j) ([raw file](https://github.com/apache/beam/blob/b8a871c118126a47dcc5d1729e3e3267c8924230/sdks/java/nexmark/src/test/java/org/apache/beam/sdk/nexmark/queries/QueryTest.java#L98)):*
> ```Java
>   @Test
>   @Category(NeedsRunner.class)
>   public void sqlQuery1MatchesModelBatch() {
> ```

Missing sqlQuery0MatchesModelStreaming and sqlQuery0MatchesModelBatch?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4718)*
<!-- Sent from Reviewable.io -->

apilloud commented:
@echauchot I've rebased and squashed commits. Unfortunately our current SQL implementation can't accurately reproduce java Query 0, as it returns multiple different types. As a result the model test will not pass so I haven't hooked it up.

apilloud commented:
run java gradle precommit

echauchot commented:
Fair enough for SQL query0 test. I merge
